api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java

```diff
@@ -740,7 +750,262 @@ public void shouldTestPurgeProgramAttributeType() throws Exception {
+			state.setState(null);
+		}
+		pws.savePatientProgram(patientProgram);
+	}
+	
+	@Test
+	public void savePatientProgram_shouldTestSetPatientProgram() {
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		for (PatientState state : patientProgram.getStates()) {
+			state.setPatientProgram(null);
+		}
+		pws.savePatientProgram(patientProgram);
+		for (PatientState state : patientProgram.getStates()) {
+			assertEquals(state.getPatientProgram(), patientProgram);
+		}
+	}
+	
+	@Rule
+	public ExpectedException expectedException2 = ExpectedException.none();
+	
+	@Test
+	public void savePatientProgram_shouldThrowPatientProgramAlreadyAssignedException() {
+		expectedException2.expect(APIException.class);
+		expectedException2.expectMessage("This PatientProgram contains a ProgramWorkflowState whose parent is already assigned " +
+											 "to PatientProgram(id=2, patient=Patient#2, program=Program(id=2, concept=Concept #10, workflows=[]))");
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		PatientProgram patientProgram1 = pws.getPatientProgram(2);
+		for (PatientState state : patientProgram.getStates()) {
+			state.setPatientProgram(patientProgram1);
+		}
+		pws.savePatientProgram(patientProgram);
+	}
+	
+	@Test
+	public void savePatientProgram_shouldTestSetState() {
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		pws.voidPatientProgram(patientProgram, "test");
+		for (PatientState state : patientProgram.getStates()) {
+			state.setVoided(false);
+			state.setVoidReason(null);
+		}
+		pws.savePatientProgram(patientProgram);
+		for (PatientState state : patientProgram.getStates()) {
+			assertEquals(state.getVoided(), true);
+			assertEquals(state.getVoidReason(), "test");
+		}
+	}
+	
+	@Test
+	public void getPrograms_shouldTestGetProgramsIfCohortIsEmpty() {
+		Cohort cohort = new Cohort();
+		Collection<Program> programs = pws.getAllPrograms();
+		List<PatientProgram> patientPrograms;
+		cohort.getMemberIds().clear();
+		patientPrograms = dao.getPatientPrograms(null, programs);
+		assertEquals(patientPrograms.size(), 4);
+	}
+	
+	@Test
+	public void getPrograms_shouldTestGetProgramsIfCohortIsNotEmpty() {
+		int i = 0;
+		Cohort cohort = new Cohort();
+		Collection<Program> programs = pws.getAllPrograms();
+		List<PatientProgram> patientPrograms;
+		cohort.addMember(1);
+		cohort.addMember(2);
+		patientPrograms = dao.getPatientPrograms(cohort, programs);
+		assertEquals(patientPrograms.size(), 2);
+	}
+	
+	
+	@Test
+	public void voidPatientProgram_shouldTestVoidPatientProgram() {
+		PatientProgram patientProgram = pws.getPatientProgram(1);
+		PatientProgram patientProgram1 = pws.voidPatientProgram(patientProgram, "abc");
+		assertEquals(patientProgram1.getVoided(), true);
+		assertEquals(patientProgram1.getVoidReason(), "abc");
+	}
+	
+	@Test
+	public void unvoidPatientProgram_shouldTestUnvoidPatientProgram() {
+		//patientProgram retains a PatientProgram
```

Comment:
It is not a good practice to write comments for each variable you write. Make your variable names to interpret what it does instead of writing comments. So, please remove those comments.

Reply:
I removed them